server/docs/design/block-verification.md

```diff
@@ -0,0 +1,198 @@
+# Block Verification Design
+
+## Table of Contents
+
+1. [Purpose](#purpose)
+2. [Goals](#goals)
+3. [Terms](#terms)
+4. [Entities](#entities)
+5. [Design](#design)
+6. [Interfaces](#interfaces)
+7. [Enums](#enums)
+8. [Metrics](#metrics)
+9. [Exceptions](#exceptions)
+
+## Purpose
+
+The purpose of the Block Verification feature is to ensure that blocks received from consensus nodes are valid and have not been tampered with. This is achieved by re-calculating the block hash and verifying it against the signature provided by the consensus node.
+
+## Goals
+
+1) The block-node must re-create the block hash from the block items and verify that it matches the hash implied by the signature.
+2) If verification fails, the block should be considered invalid, and appropriate error-handling procedures must be triggered.
+
+
+## Terms
+
+- Consensus Node (CN): A node that produces and provides blocks.
+- Block Items: The block data pieces (header, events, transactions, transaction result, state changes, proof) that make up a block.
+- Block Hash: A cryptographic hash representing the block’s integrity.
+- Signature: The cryptographic signature of the block hash created by Network private key (aka: LedgerId)
+- Public Key: The public key of the LedgerId that signed the block.
+
+
+## Entities
+
+- **VerificationHandler** 
+  - Receives the stream of block items in the form of List<BlockItemsUnparsed> from the unparsed and unverified block items ring buffer. 
+  - When it detects a block_header, it creates a BlockHashingSession using the BlockHashingSessionFactory, providing it with the initial block items (and internally, the session will handle asynchronous hashing).
+  - Adds subsequent block items to the session, including the block_proof.
+  - Once the block_proof is received, calls completeHashing() on the BlockHashingSession, then releases the thread back to the UnverifiedRingBuffer.
+  - Does not block waiting for verification; the hash computation and verification continue asynchronously.
+- **BlockHashingSessionFactory**
+  - Creates new BlockHashingSession instances, provides them with a ExecutorService.
+- **BlockHashingSession**
+  - Holds all necessary block data.
+  - Accepts block items incrementally. (continues to compute them asynchronously)
+  - Once the block_proof is provided, finalizes the hash computation asynchronously.
+  - After computing the final hash, calls SignatureVerifier for verification.
+- **SignatureVerifier**
+  - Verifies the signature by comparing the computed hash to the hash implied by the signature (using the public key).
+  - Calls BlockStatusManager with the verification result.
+- **BlockStatusManager**
+  - Receives verification results from SignatureVerifier.
+  - Updates block status and triggers any necessary recovery or follow-up processes depending on the outcome.
```

Comment:
I guess that we could create a "Verified Ring Buffer" that the SignatureVerifier writes the block there. maybe even already parsed, so that the `StateHandler` can pick it up from there. instead of relying on a coordinator/manager.

However how things are working now, the server is returning `ack` messages to the `producer` and it knows when to return them using another manager `Notifier` right now the Notifier returns the ack to the producer when the block finishes persistence, however it should be done once the block is verified, so my short term idea for this solution is to have a BlockManager that basically holds the last few blocks in memory, and gets updates of the blocks as it goes throught the pipelines (persistence, verification, stateUpdater, etc...) the BlockManager not only keeps context of `work` that was performed on the last few blocks such as `parsing`, `verification`, but also orchestrates the workflow (journey of a block_items)

Reply:
In this case the intent of the design doc is to communicate the idea of what we intent to do in the short and immediate future, **current Verification Epic**. and once we settle on a new design for long term we should make part of that effort to update the corresponding docs. I have captured the ideas coming off this PR review in a separate [discussion](https://github.com/hashgraph/hedera-block-node/discussions/400) that we can create an Epic from.

Reply:
Fair.  We should still be careful not to paint ourselves in a corner in this document, but that's a judgement call.